File: Microsoft.CodeAnalysis.VisualBasic/Microsoft.CodeAnalysis.VisualBasic.Syntax/ElseIfStatementSyntax.cs
using Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax;

namespace Microsoft.CodeAnalysis.VisualBasic.Syntax
{
	public sealed class ElseIfStatementSyntax : StatementSyntax
	{
		internal ExpressionSyntax _condition;

		public SyntaxToken ElseIfKeyword => new SyntaxToken(this, ((Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.ElseIfStatementSyntax)base.Green)._elseIfKeyword, base.Position, 0);

		public ExpressionSyntax Condition => GetRed(ref _condition, 1);

		public SyntaxToken ThenKeyword
		{
			get
			{
				KeywordSyntax thenKeyword = ((Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.ElseIfStatementSyntax)base.Green)._thenKeyword;
				return (thenKeyword == null) ? default(SyntaxToken) : new SyntaxToken(this, thenKeyword, GetChildPosition(2), GetChildIndex(2));
			}
		}

		internal ElseIfStatementSyntax(GreenNode green, SyntaxNode parent, int startLocation)
			: base(green, parent, startLocation)
		{
		}

		internal ElseIfStatementSyntax(SyntaxKind kind, DiagnosticInfo[] errors, SyntaxAnnotation[] annotations, KeywordSyntax elseIfKeyword, ExpressionSyntax condition, KeywordSyntax thenKeyword)
			: this(new Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.ElseIfStatementSyntax(kind, errors, annotations, elseIfKeyword, (Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.ExpressionSyntax)condition.Green, thenKeyword), null, 0)
		{
		}

		public ElseIfStatementSyntax WithElseIfKeyword(SyntaxToken elseIfKeyword)
		{
			return Update(elseIfKeyword, Condition, ThenKeyword);
		}

		public ElseIfStatementSyntax WithCondition(ExpressionSyntax condition)
		{
			return Update(ElseIfKeyword, condition, ThenKeyword);
		}

		public ElseIfStatementSyntax WithThenKeyword(SyntaxToken thenKeyword)
		{
			return Update(ElseIfKeyword, Condition, thenKeyword);
		}

		internal override SyntaxNode GetCachedSlot(int i)
		{
			if (i == 1)
			{
				return _condition;
			}
			return null;
		}

		internal override SyntaxNode GetNodeSlot(int i)
		{
			if (i == 1)
			{
				return Condition;
			}
			return null;
		}

		public override TResult Accept<TResult>(VisualBasicSyntaxVisitor<TResult> visitor)
		{
			return visitor.VisitElseIfStatement(this);
		}

		public override void Accept(VisualBasicSyntaxVisitor visitor)
		{
			visitor.VisitElseIfStatement(this);
		}

		public ElseIfStatementSyntax Update(SyntaxToken elseIfKeyword, ExpressionSyntax condition, SyntaxToken thenKeyword)
		{
			if (elseIfKeyword != ElseIfKeyword || condition != Condition || thenKeyword != ThenKeyword)
			{
				ElseIfStatementSyntax elseIfStatementSyntax = SyntaxFactory.ElseIfStatement(elseIfKeyword, condition, thenKeyword);
				SyntaxAnnotation[] annotations = GetAnnotations();
				if (annotations != null && annotations.Length > 0)
				{
					return SyntaxNodeExtensions.WithAnnotations(elseIfStatementSyntax, annotations);
				}
				return elseIfStatementSyntax;
			}
			return this;
		}
	}
}

File: Microsoft.CodeAnalysis.VisualBasic/Microsoft.CodeAnalysis/VisualBasic/Syntax/ElseIfStatementSyntax.vb
Imports Microsoft.CodeAnalysis
Imports Microsoft.CodeAnalysis.VisualBasic
Imports Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax
Imports System

Namespace Microsoft.CodeAnalysis.VisualBasic.Syntax
	Public NotInheritable Class ElseIfStatementSyntax
		Inherits Microsoft.CodeAnalysis.VisualBasic.Syntax.StatementSyntax
		Friend _condition As Microsoft.CodeAnalysis.VisualBasic.Syntax.ExpressionSyntax

		Public ReadOnly Property Condition As Microsoft.CodeAnalysis.VisualBasic.Syntax.ExpressionSyntax
			Get
				Return MyBase.GetRed(Of Microsoft.CodeAnalysis.VisualBasic.Syntax.ExpressionSyntax)(Me._condition, 1)
			End Get
		End Property

		Public ReadOnly Property ElseIfKeyword As Microsoft.CodeAnalysis.SyntaxToken
			Get
				Return New Microsoft.CodeAnalysis.SyntaxToken(Me, DirectCast(MyBase.Green, Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.ElseIfStatementSyntax)._elseIfKeyword, MyBase.Position, 0)
			End Get
		End Property

		Public ReadOnly Property ThenKeyword As Microsoft.CodeAnalysis.SyntaxToken
			Get
				Dim syntaxToken As Microsoft.CodeAnalysis.SyntaxToken
				Dim green As KeywordSyntax = DirectCast(MyBase.Green, Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.ElseIfStatementSyntax)._thenKeyword
				syntaxToken = If(green Is Nothing, New Microsoft.CodeAnalysis.SyntaxToken(), New Microsoft.CodeAnalysis.SyntaxToken(Me, green, Me.GetChildPosition(2), MyBase.GetChildIndex(2)))
				Return syntaxToken
			End Get
		End Property

		Friend Sub New(ByVal green As GreenNode, ByVal parent As SyntaxNode, ByVal startLocation As Integer)
			MyBase.New(green, parent, startLocation)
		End Sub

		Friend Sub New(ByVal kind As SyntaxKind, ByVal errors As DiagnosticInfo(), ByVal annotations As SyntaxAnnotation(), ByVal elseIfKeyword As KeywordSyntax, ByVal condition As Microsoft.CodeAnalysis.VisualBasic.Syntax.ExpressionSyntax, ByVal thenKeyword As KeywordSyntax)
			MyClass.New(New Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.ElseIfStatementSyntax(kind, errors, annotations, elseIfKeyword, DirectCast(condition.Green, Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.ExpressionSyntax), thenKeyword), Nothing, 0)
		End Sub

		Public Overrides Function Accept(Of TResult)(ByVal visitor As VisualBasicSyntaxVisitor(Of TResult)) As TResult
			Return visitor.VisitElseIfStatement(Me)
		End Function

		Public Overrides Sub Accept(ByVal visitor As Microsoft.CodeAnalysis.VisualBasic.VisualBasicSyntaxVisitor)
			visitor.VisitElseIfStatement(Me)
		End Sub

		Friend Overrides Function GetCachedSlot(ByVal i As Integer) As Microsoft.CodeAnalysis.SyntaxNode
			Dim syntaxNode As Microsoft.CodeAnalysis.SyntaxNode
			If (i <> 1) Then
				syntaxNode = Nothing
			Else
				syntaxNode = Me._condition
			End If
			Return syntaxNode
		End Function

		Friend Overrides Function GetNodeSlot(ByVal i As Integer) As SyntaxNode
			Dim condition As SyntaxNode
			If (i <> 1) Then
				condition = Nothing
			Else
				condition = Me.Condition
			End If
			Return condition
		End Function

		Public Function Update(ByVal elseIfKeyword As Microsoft.CodeAnalysis.SyntaxToken, ByVal condition As Microsoft.CodeAnalysis.VisualBasic.Syntax.ExpressionSyntax, ByVal thenKeyword As Microsoft.CodeAnalysis.SyntaxToken) As Microsoft.CodeAnalysis.VisualBasic.Syntax.ElseIfStatementSyntax
			Dim elseIfStatementSyntax As Microsoft.CodeAnalysis.VisualBasic.Syntax.ElseIfStatementSyntax
			If (elseIfKeyword <> Me.ElseIfKeyword OrElse condition <> Me.Condition OrElse thenKeyword <> Me.ThenKeyword) Then
				Dim elseIfStatementSyntax1 As Microsoft.CodeAnalysis.VisualBasic.Syntax.ElseIfStatementSyntax = Microsoft.CodeAnalysis.VisualBasic.SyntaxFactory.ElseIfStatement(elseIfKeyword, condition, thenKeyword)
				Dim annotations As SyntaxAnnotation() = MyBase.GetAnnotations()
				elseIfStatementSyntax = If(annotations Is Nothing OrElse CInt(annotations.Length) <= 0, elseIfStatementSyntax1, elseIfStatementSyntax1.WithAnnotations(annotations))
			Else
				elseIfStatementSyntax = Me
			End If
			Return elseIfStatementSyntax
		End Function

		Public Function WithCondition(ByVal condition As Microsoft.CodeAnalysis.VisualBasic.Syntax.ExpressionSyntax) As Microsoft.CodeAnalysis.VisualBasic.Syntax.ElseIfStatementSyntax
			Return Me.Update(Me.ElseIfKeyword, condition, Me.ThenKeyword)
		End Function

		Public Function WithElseIfKeyword(ByVal elseIfKeyword As Microsoft.CodeAnalysis.SyntaxToken) As Microsoft.CodeAnalysis.VisualBasic.Syntax.ElseIfStatementSyntax
			Return Me.Update(elseIfKeyword, Me.Condition, Me.ThenKeyword)
		End Function

		Public Function WithThenKeyword(ByVal thenKeyword As Microsoft.CodeAnalysis.SyntaxToken) As Microsoft.CodeAnalysis.VisualBasic.Syntax.ElseIfStatementSyntax
			Return Me.Update(Me.ElseIfKeyword, Me.Condition, thenKeyword)
		End Function
	End Class
End Namespace